lib/task/jobs/jobWorkerTask.class.php

```diff
@@ -120,14 +165,33 @@ function ($handle, $job, $e) {
             // Another option would be to catch the ProperException from the worker
             // and restablish the connection when needed. Also, the persistent mode
             // could be disabled for this worker. See issue #4182.
-            function () use (&$counter) {
+            function ($idle, $lastJob) use (&$counter) {
+                if ($this->maxJobCountReached()) {
+                    $this->log(sprintf('Max job count reached: %u jobs completed.', $this->getJobsCompleted()), sfLogger::INFO);
+
+                    // Notify the worker that beginWork() work loop should exit.
+                    return true;
+                }
+
                 if (30 == $counter++) {
                     $counter = 0;
 
                     QubitPdo::prepareAndExecute('SELECT 1');
                 }
             }
         );
+
+        // Return code '111' when worker shuts down due to reaching max job count.
+        if ($this->getMaxJobCount() > 0 && $this->maxJobCountReached()) {
+            $this->log('Worker shutting down - max-job-count reached.');
+            // Force worker's __destruct() to run so gearman connection is closed nicely.
+            unset($worker);
```

Comment:
I don't think there's any advantage to `unset($worker);` before `exit()` - all the memory should be freed when the script is killed, AFAIK.

Reply:
The same comment applies here as with the second call to `unset($worker)` below. This forces the worker's destructor to run which triggers a clean disconnect from Gearman.

https://github.com/artefactual/atom/blob/qa/2.x/vendor/net_gearman/Net/Gearman/Worker.php#L456-L472

Without this line, debug statements indicate that the worker's destructor is not run on exit() - I'd rather have the worker disconnect 'nicely' from gearman to avoid issues with open ports etc.

Reply:
Okay, this makes sense to me. 👍